source/common/upstream/load_balancer_impl.cc

```diff
@@ -88,6 +99,11 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
     per_priority_load_[i] =
         std::min<uint32_t>(total_load, per_priority_health_[i] * 100 / total_health);
     total_load -= per_priority_load_[i];
+
+    // For each level check if it should run in Panic mode. Never set Panic mode if the total health
+    // is 100%, even when individual priority level has very low # of healthy hosts.
+    const HostSet& priority_host_set = *priority_set_.hostSetsPerPriority()[i];
+    per_priority_panic_[i] = (total_health == 100 ? false : isGlobalPanic(priority_host_set));
```

Comment:
As a reader, I wonder if everywhere here we talk about "total health" we really should call it something else like "normalized total health" or something like that? It's a little confusing as we apply a min function over all priorities. Thoughts? Mostly just trying to make this very complicated code easier to understand for the next person to come along. cc @alyssawilk and @htuch also who have worked in this area.

Reply:
Yeaaaah, I encourage making it easier to understand but I don't think I have better advice than yours :-/